### PR TITLE
Support downloading raw task logs

### PR DESCRIPTION
**Pull request by vitalif**
Hi!
This pull request adds support for downloading raw task logs for Gitea Actions, similar to Github Actions
It looks like the following:
image

### CODE REVIEW

File: routers/web/repo/actions/view.go [hunk 1]
```diff
@@ -270,6 +271,54 @@ func Rerun(ctx *context_module.Context) {
...
+		Filename: job.Name+".txt",
+		ContentLength: &task.LogSize,
+		ContentType: "text/plain",
+		ContentTypeCharset: "utf-8",
+		Disposition: "attachment",
+	}
+
+	ctx.SetServeHeaders(opts)
+
+	_, err = io.Copy(ctx.Resp, reader)
```
Comment (wxiaoguang) [inline routers/web/repo/actions/view.go]: `ctx.ServeContent` ?

**Comment by yp05327** [comment]
What about the permission? Anyone who can access this page can download them?

File: web_src/js/components/RepoActionView.vue [hunk 2]
```diff
@@ -43,11 +43,18 @@
...
+              {{ currentJob.title }}
+            </div>
+            <div class="job-info-header-detail">
+              {{ currentJob.detail }}
+            </div>
           </div>
-          <div class="job-info-header-detail">
-            {{ currentJob.detail }}
+          <div class="job-info-header-props">
+            <a class="job-brief-link" :href="run.link+'/jobs/'+jobIndex+'/logs'" target="_blank">
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue]:
```suggestion
            <a class="job-brief-link muted" :href="run.link+'/jobs/'+jobIndex+'/logs'">
```

- `muted` removes color unless hovered.
- `target` is unnecessary when server serves as `attachement`.

File: web_src/js/components/RepoActionView.vue [hunk 3]
```diff
@@ -469,14 +476,28 @@ export function ansiLogToHTML(line) {
   flex-direction: column;
 }
 
+.job-info-header {
+  padding: 10px;
+  border-bottom: 1px solid var(--color-grey);
```
Comment (silverwind) [inline web_src/js/components/RepoActionView.vue:481]:
```suggestion
  border-bottom: 1px solid var(--color-secondary);
```

Should look pretty similar, but imho it's better we use this variable.

**Comment by wolfogre** [comment]
> What about the permission? Anyone who can access this page can download them?

IMO, yes. Since one can access this page, they should be able to see the log content on the page. So why can't they download the log file?

**Comment by yp05327** [comment]
> > What about the permission? Anyone who can access this page can download them?
> 
> IMO, yes. Since one can access this page, they should be able to see the log content on the page. So why can't they download the log file?

Do these logs have secrets?
I’m not sure where we convert secrets into stars.

**Comment by wolfogre** [comment]
> > > What about the permission? Anyone who can access this page can download them?
> > 
> > 
> > IMO, yes. Since one can access this page, they should be able to see the log content on the page. So why can't they download the log file?
> 
> Do these logs have secrets? I’m not sure where we convert secrets into ‘*’

The secrets in logs will be converted into `***` by runners before the logs are sent to Gitea.

File: routers/web/repo/actions/view.go (hunk 1 above)
Comment (vitalif) [reply routers/web/repo/actions/view.go]: Thanks, changed

File: web_src/js/components/RepoActionView.vue (hunk 2 above)
Comment (vitalif) [reply web_src/js/components/RepoActionView.vue]: Thanks, but I see there's a menu in the rebased version, so I added it as a menu item

File: routers/web/repo/actions/view.go [hunk 4]
```diff
@@ -310,6 +310,46 @@ func rerunJob(ctx *context_module.Context, job *actions_model.ActionRunJob) erro
...
 
+func Logs(ctx *context_module.Context) {
+	runIndex := ctx.ParamsInt64("run")
+	jobIndex := ctx.ParamsInt64("job")
+
+	job, _ := getRunJobs(ctx, runIndex, jobIndex)
+	if ctx.Written() {
+		return
+	}
+	if !job.Status.HasRun() && !job.Status.IsRunning() || job.TaskID == 0 {
```
Comment (wolfogre) [inline routers/web/repo/actions/view.go]: I think
```suggestion
	if job.TaskID == 0 {
```
should be enough.

File: routers/web/repo/actions/view.go [hunk 5]
```diff
@@ -310,6 +310,46 @@ func rerunJob(ctx *context_module.Context, job *actions_model.ActionRunJob) erro
...
+	task, err := actions_model.GetTaskByID(ctx, job.TaskID)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+	task.Job = job
+	if err := task.LoadAttributes(ctx); err != nil {
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
```
Comment (wolfogre) [inline routers/web/repo/actions/view.go]: Maybe it's unnecessary:
```suggestion
```

It doesn't use `task.Job` or `task.Steps` here.

Comment (vitalif) [reply routers/web/repo/actions/view.go]: Yes it seems you're right. Thanks

File: routers/web/repo/actions/view.go (hunk 4 above)
Comment (vitalif) [reply routers/web/repo/actions/view.go]: Ok, changed it too

File: web_src/js/components/RepoActionView.vue (hunk 2 above)
Comment (vitalif) [reply web_src/js/components/RepoActionView.vue]: I re-added `target="_blank"` because it prevents you from losing your tab if the server returns error

Comment (silverwind) [reply web_src/js/components/RepoActionView.vue]: I guess you want to instead set the [`download`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#download) attribute then, likely in boolean form. `target=_blank` is not right for downloads, this one is, e.g.

```html
<a download href="foo"></a>

File: routers/web/repo/actions/view.go [hunk 6]
```diff
@@ -310,6 +310,41 @@ func rerunJob(ctx *context_module.Context, job *actions_model.ActionRunJob) erro
...
+		ctx.Error(http.StatusNotFound, "job is not started")
+		return
+	}
+
+	task, err := actions_model.GetTaskByID(ctx, job.TaskID)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
```
Comment (wolfogre) [inline routers/web/repo/actions/view.go:342]:
```suggestion
	if task.LogExpired {
		ctx.Error(http.StatusNotFound, "logs have been cleaned up")
		return
	}

```

**Comment by wolfogre** [comment]
I have no more questions on the backend side, it works well on my environment. But I think there's still some room for improvement on the UI.

- It doen't look well:
image

- If it's unavailable to download, I can still click the button and get a text error page.
  - Maybe the button should be disabled, and show a tooltip to explain why.
image
image

Not blocking, it doesn't really hurt. I'm OK to merge this first and improve it in another PR.

File: routers/web/repo/actions/view.go [hunk 7]
```diff
@@ -310,6 +310,41 @@ func rerunJob(ctx *context_module.Context, job *actions_model.ActionRunJob) erro
...
+	reader, err := actions.OpenLogs(ctx, task.LogInStorage, task.LogFilename)
+	if err != nil {
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+	defer reader.Close()
+
+	ctx.ServeContent(reader, &context_module.ServeHeaderOptions{
+		Filename:           job.Name + ".txt",
```
Comment (lunny) [inline routers/web/repo/actions/view.go]: .log? Maybe the name should include the number.

**Comment by silverwind** [comment]
This alignment issue is from Fomantic UI. We could circumvent it by adding icons to all menu items. In case of "unchecked" state, we could create a `gitea-empty.svg` which is a SVG without any content, so it takes up space.

File: routers/web/repo/actions/view.go (hunk 7 above)
Comment (silverwind) [reply routers/web/repo/actions/view.go]: I would do a pattern like `$workflow-$job-$id.log`.

File: web_src/js/components/RepoActionView.vue (hunk 2 above)
Comment (silverwind) [reply web_src/js/components/RepoActionView.vue]: Done that in https://github.com/go-gitea/gitea/pull/24451/commits/5e29e0592b51f39742e1c757e50e5a5fd6da3860, now there's no more flash of a tab opening and closing.

**Comment by silverwind** [comment]
> 
> 
> This alignment issue can be blamed on Fomantic UI. We could circumvent it by adding icons to all menu items. In case of "unchecked" state, we could create a `gitea-empty.svg` which is a SVG without any content, so it takes up space.

Fixed in https://github.com/go-gitea/gitea/pull/24451/commits/228bc9bf8d1212a4aa451478ce9e8fb27e5c0183:

Screenshot 2023-06-06 at 20 11 37

**Review by techknowlogick — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Comment by vitalif** [comment]
Oh, thanks for merging, I was thinking about fixing UI :)

**Comment by silverwind** [comment]
Still fixes outstanding?

**Comment by vitalif** [comment]
Ok, no problem :)

**Comment by yp05327** [comment]
> Done with my fixes, leaving the rest to you @vitalif.

@silverwind 
It seems that your fix in https://github.com/go-gitea/gitea/commit/228bc9bf8d1212a4aa451478ce9e8fb27e5c0183 was broken by the force push here:
https://github.com/go-gitea/gitea/compare/ae9b3e969c843f6ac9025f5f080b67af907d4a24..aeb7759f88ad2a5951322d6b8691e9cbd1b0ca5e#diff-c3422631a14edbe1e508c4b22f0c718db318be08a6e889427802f9b6165d88d6R82

**Comment by silverwind** [comment]
I don't understand. Do we need to re-apply it in a separate PR?

**Comment by yp05327** [comment]
> I don't understand. Do we need to re-apply it in a separate PR?

@silverwind 
The UI now is as the following and it looks strange:
image
Then I notice that you have improved it in https://github.com/go-gitea/gitea/commit/228bc9bf8d1212a4aa451478ce9e8fb27e5c0183
And it looks better:
image

The reason why we lost your patch is that after you push your code, it seems that @vitalif didn't notice that and made a force push to the branch:
image
And in this force push, your patch is recovered. So I think we should re-apply it in a separate PR.

**Comment by silverwind** [comment]
Indeed, the force-push eliminated my changes. Now this is rather hard to recover as all server-side data was destroyed. Maybe I still have it in my local reflog, need to check.

**Comment by silverwind** [comment]
https://github.com/go-gitea/gitea/pull/26278

**Pull request merged**